### PR TITLE
K8SPXC-167 Add user manager

### PR DESCRIPTION
**Pull request by nonemax**

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| init-deploy | failed |
| self-healing | failed |
| big-data | passed |
| scaling | passed |
| storage | failed |
| self-healing-advanced | passed |
| demand-backup | passed |
| scaling-proxysql | failed |
| limits | passed |
| operator-self-healing | passed |
| upgrade | failed |
| monitoring | failed |
| upgrade-consistency | passed |
| one-pod | passed |
| auto-tuning | passed |
| scheduled-backup | failed |
| monitoring-2-0 | failed |
| affinity | passed |
| proxysql-sidecar-res-limits | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| self-healing | failed |
| recreate | failed |
| scaling | failed |
| init-deploy | failed |
| big-data | failed |
| self-healing-advanced | failed |
| demand-backup | failed |
| scaling-proxysql | failed |
| storage | failed |
| operator-self-healing | failed |
| limits | passed |
| upgrade | failed |
| scheduled-backup | failed |
| monitoring | failed |
| one-pod | passed |
| upgrade-consistency | passed |
| auto-tuning | failed |
| monitoring-2-0 | failed |
| affinity | passed |
| proxysql-sidecar-res-limits | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| recreate | failed |
| scaling | failed |
| self-healing | failed |
| big-data | failed |
| demand-backup | failed |
| scaling-proxysql | failed |
| storage | failed |
| self-healing-advanced | failed |
| limits | passed |
| monitoring | failed |
| upgrade | failed |
| operator-self-healing | failed |
| scheduled-backup | failed |
| upgrade-consistency | passed |
| one-pod | passed |
| auto-tuning | failed |
| monitoring-2-0 | failed |
| affinity | passed |
| proxysql-sidecar-res-limits | failed |

File: pkg/controller/pxc/users.go [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+	}
+	return nil
+
+}
+
+func usersDataChanged(newHash *string, usersSecret *corev1.Secret) bool {
+	if secretData, ok := usersSecret.Data["grants.yaml"]; ok {
+		hash := sha256Hash(secretData)
+		*newHash = hash
+		if lastAppliedHash, ok := usersSecret.Annotations["last-applied"]; ok {
```
Comment (delgod) [inline pkg/controller/pxc/users.go:230]: can we remove this `if` and rename function into something like `getSecretHash`?
goal: improve readability in `reconcileUsers` func

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | failed |
| scaling | failed |
| init-deploy | failed |
| self-healing | failed |
| recreate | failed |
| scaling-proxysql | failed |
| self-healing-advanced | failed |
| storage | failed |
| demand-backup | failed |
| upgrade | failed |
| limits | passed |
| operator-self-healing | failed |
| scheduled-backup | failed |
| monitoring | failed |
| upgrade-consistency | passed |
| one-pod | passed |
| monitoring-2-0 | failed |
| auto-tuning | failed |
| affinity | passed |
| proxysql-sidecar-res-limits | failed |

File: pkg/controller/pxc/users.go (hunk 1 above)
Comment (nonemax) [reply pkg/controller/pxc/users.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/392/commits/6152288c13c92aa90e99f0c829dfa64f325e87c4

File: pkg/controller/pxc/users.go [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+	err = r.client.Update(context.TODO(), &secretObj)
+	if err != nil {
+		return errors.Wrap(err, "create internal users secret")
+	}
+	return nil
+
+}
+
+func usersDataChanged(newHash *string, usersSecret *corev1.Secret) bool {
+	if secretData, ok := usersSecret.Data["grants.yaml"]; ok {
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/controller/pxc/users.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/392/commits/f8dd2bb2fa2ec8f14add6896d20fdb9168cfa55f

File: pkg/apis/pxc/v1/pxc_types.go [hunk 3]
```diff
@@ -206,6 +207,20 @@ type BackupStorageS3Spec struct {
...
+	NodeSelector             map[string]string          `json:"nodeSelector,omitempty"`
+	Resources                *PodResources              `json:"resources,omitempty"`
+	Affinity                 *corev1.Affinity           `json:"affinity,omitempty"`
+	Tolerations              []corev1.Toleration        `json:"tolerations,omitempty"`
+	Annotations              map[string]string          `json:"annotations,omitempty"`
+	Labels                   map[string]string          `json:"labels,omitempty"`
+	SchedulerName            string                     `json:"schedulerName,omitempty"`
+	PriorityClassName        string                     `json:"priorityClassName,omitempty"`
+	PodSecurityContext       *corev1.PodSecurityContext `json:"podSecurityContext,omitempty"`
+	ContainerSecurityContext *corev1.SecurityContext    `json:"containerSecurityContext,omitempty"`
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/apis/pxc/v1/pxc_types.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/392/commits/3853c1c7a27084a4d61d47d2feaaa98cd71e1ccb

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| self-healing | failed |
| big-data | failed |
| recreate | failed |
| scaling | failed |
| init-deploy | failed |
| demand-backup | failed |
| self-healing-advanced | failed |
| scaling-proxysql | failed |
| storage | failed |
| operator-self-healing | failed |
| upgrade | failed |
| limits | passed |
| scheduled-backup | failed |
| monitoring | failed |
| one-pod | passed |
| upgrade-consistency | passed |
| auto-tuning | failed |
| monitoring-2-0 | failed |
| affinity | passed |
| proxysql-sidecar-res-limits | failed |

**Comment by delgod** [comment]
archived due to roadmap changes. intent to reopen later this year

**Pull request closed without merging**